core/src/main/java/org/apache/spark/shuffle/api/ShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.api;
+
+import java.io.IOException;
+
+import org.apache.spark.annotation.Private;
+
+/**
+ * :: Experimental ::
+ * A top-level writer that returns child writers for persisting the output of a map task,
+ * and then commits all of the writes as one atomic operation.
+ *
+ * @since 3.0.0
+ */
+@Private
+public interface ShuffleMapOutputWriter {
+
+  /**
+   * Creates a writer that can open an output stream to persist bytes for a given chunk of
+   * a map task.
+   * <p>
+   * The chunk corresponds to bytes in a partition that all share the same reduce id, hence
+   * the given argument. This will not be called twice for the same partition identifier.
+   * The partition identifier will be in the range of precisely 0 (inclusive) to numPartitions
+   * (exclusive), where numPartitions was provided upon the creation of this map output writer via
+   * {@link ShuffleWriteSupport#createMapOutputWriter(int, int, long, int)}.
+   */
+  ShufflePartitionWriter getPartitionWriter(int partitionId) throws IOException;
+
+  /**
+   * Commits the writes done by all partition writers returned by all calls to this object's
+   * {@link #getPartitionWriter(int)}.
+   * <p>
+   * This should ensure that the writes conducted by this module's partition writers are
+   * available to downstream reduce tasks. If this method throws any exception, this module's
+   * {@link #abort(Throwable)} method will be invoked before propagating the exception.
+   * <p>
+   * This can also close any resources and clean up temporary state if necessary.
+   */
+  void commitAllPartitions() throws IOException;
```

Comment:
Shouldn't this return `Optional<MapShuffleLocations>`?

Reply:
@gczsjdy any reason to return `Optional<MapShuffleLocations>`?

Reply:
We ended up adjusting the API for shuffle locations. This will come later.

Reply:
I believe the SPIP has the latest API.

Reply:
@jerryshao @mccheah has explained well, because `Optional<MapShuffleLocations>` make implementers customize locations recorded in Driver. @mccheah This will be in driver lifecycle subissue?

Reply:
Something to that effect yeah - it also has implications on the reader API, but these are concerns to be addressed in subsequent patches.

Reply:
Got it. : )